extensions/webpush.md

```diff
@@ -0,0 +1,96 @@
+
+To address these limitations, various push notification mechanisms have been designed. This specification standardizes an extension for Web Push.
+
+```
+      ┌────────────┐              ┌────────────┐
+      │            │  Subscribe   │            │
+      │            ├─────────────►│            │
+      │ IRC client │              │ IRC server │
+      │            │              │            │
+      │            │              │            │
+      └────────────┘              └─────┬──────┘
+             ▲                          │
+             │                          │
+        Push │                          │Push
+notification │                          │notification
+             │       ┌──────────┐       │
+             │       │          │       │
+             └───────┤ Web Push │◄──────┘
+                     │  Server  │
+                     │          │
+                     └──────────┘
+```
+
+Web Push is defined in [RFC 8030], [RFC 8291] and [RFC 8292].
+
+Although Web Push has been designed for the Web, it can be used on other platforms as well. Web Push provides a vendor-neutral standard to send push notifications.
+
+## Implementation
+
+The `webpush` capability allows clients to subscribe to Web Push and receive
+notifications for messages of interest.
+
+Once a client has subscribed, the server will send push notifications for a
+server-defined subset of IRC messages. Each push notification MUST contain
+exactly one IRC message as the payload, without the final CRLF.
+
+## `WEBPUSH` Command
+
+A new `WEBPUSH` command is introduced. It has a case-insensitive subcommand:
+
+    WEBPUSH <subcommand> <params...>
+
+### `VAPIDPUBKEY` Subcommand
+
+The `VAPIDPUBKEY` subcommand allows clients to query the [Voluntary Application Server Identification (VAPID)][RFC 8292] public key of the server. This can be used to decrypt notifications upon reception.
+
+The client can query the VAPID public key by sending the command:
+
+    WEBPUSH VAPIDPUBKEY
+
+The server will reply with a [URL-safe base64-encoded][RFC 4648 section 5] public key usable with the Elliptic Curve Digital Signature Algorithm (ECDSA) over the P-256 curve.
+
+    WEBPUSH VAPIDPUBKEY <key>
+
+### `REGISTER` Subcommand
+
+The `REGISTER` subcommand creates a new Web Push subscription.
+
+    WEBPUSH REGISTER <endpoint> <keys>
+
+The `<endpoint>` is an URL pointing to a push server, which can be used to send push messages for this particular subscription.
+
+`<keys>` is a string encoded in the message-tag format. The values are [URL-safe base64-encoded][RFC 4648 section 5]. It MUST contain at least:
+
+- One public key with the name `p256dh` set to the client's P-256 ECDH Diffie-Hellman public key.
+- One shared key with the name `auth` set to a client-generated authentication secret.
```

Comment:
I would recommend wording the spec a little differently so that algorithm(s) can be updated over time. Ensuring the design of the protocol is easy to update based on discovered weaknesses or when stronger algorithms come around. Prevent using terms like "these fields are required", but "for the algorithm X these fields are required".

Reply:
This part follows exactly the Web Push specs, which makes these two required.

Reply:
The set of specifications surrounding web push are separated into individual RFCs. RFC8030 does not mandate encryption nor an single encryption standard. 

Whereas here we have one specification that defines both the web push capabilities, how to negotiate clients which requires one specific encryption mechanisms. This doesn't appear to be the case for web push.

RFC 8030 "Generic Event Delivery Using HTTP Push" specifies:

> Examples do not include specific methods for push message encryption or application server authentication because the protocol does not define a mandatory system.  The examples in Voluntary Application Server Identification [VAPID] and Message Encryption for WebPush [ENCRYPT] demonstrate the approach adopted by the W3C Push API [API] for its requirements.

[RFC8291](https://datatracker.ietf.org/doc/html/rfc8291) on its own only describes how to use this particular algorithm and does not describe how web push works. This RFC can be replaced on its own.

This is similar to how IRCv3 has specified SASL, the mechanisms are separate specs which live in their own lifecycle from core SASL.

Reply:
I see, thanks for taking the time to explain. I'm not sure how to word this. Would this do it? "For the encryption scheme described in RFC 8291, these fields are required"